crates/spirv-std/src/arch.rs

```diff
@@ -150,3 +150,25 @@ pub unsafe fn vector_insert_dynamic<T: Scalar, V: Vector<T, N>, const N: usize>(
 pub fn kill() -> ! {
     unsafe { asm!("OpKill", options(noreturn)) }
 }
+
+/// Read from the shader clock.
+///
+/// Requires the `SPV_KHR_shader_clock` extension and the `ShaderClockKHR` capability.
+///
+/// See:
+/// <https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_shader_clock.html>
+#[spirv_std_macros::gpu_only]
+#[doc(alias = "OpReadClockKHR")]
+pub unsafe fn read_clock_khr() -> u64 {
+    let mut result: u64 = 0;
+
+    asm! {
+        "%uint = OpTypeInt 32 0",
+        "%uint_3 = OpConstant %uint 3",
```

Comment:
I'm not entirely sure how to do this 'scope' field. It has to be an `OpConstant` so I'm not sure if that's something we can use `in(reg)` for. 

With glsl:

```glsl
#version 450
#extension GL_EXT_shader_explicit_arithmetic_types_int64 : enable
#extension GL_ARB_shader_clock : enable

void main() {
	uint64_t clock_value = clockARB();
}
```

becomes:

```spirv
; SPIR-V
; Version: 1.0
; Generator: Google Shaderc over Glslang; 10
; Bound: 12
; Schema: 0
               OpCapability Shader
               OpCapability Int64
               OpCapability ShaderClockKHR
               OpExtension "SPV_KHR_shader_clock"
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main"
               OpExecutionMode %main OriginUpperLeft
               OpSource GLSL 450
               OpSourceExtension "GL_ARB_shader_clock"
               OpSourceExtension "GL_EXT_shader_explicit_arithmetic_types_int64"
               OpSourceExtension "GL_GOOGLE_cpp_style_line_directive"
               OpSourceExtension "GL_GOOGLE_include_directive"
               OpName %main "main"
               OpName %clock_value "clock_value"
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
      %ulong = OpTypeInt 64 0
%_ptr_Function_ulong = OpTypePointer Function %ulong
       %uint = OpTypeInt 32 0
     %uint_3 = OpConstant %uint 3
       %main = OpFunction %void None %3
          %5 = OpLabel
%clock_value = OpVariable %_ptr_Function_ulong Function
         %11 = OpReadClockKHR %ulong %uint_3
               OpStore %clock_value %11
               OpReturn
               OpFunctionEnd
```

so I just copied that constant.

Reply:
Using `const` in combination with the `Scope` enum from `memory` would be an option:
```rust
let mut result: u64;

asm! {
    "%uint = OpTypeInt 32 0",
    "%scope = OpConstant %uint {scope}",
    "{result} = OpReadClockKHR typeof*{result} %scope",
    result = out(reg) result,
    scope = const Scope::Subgroup as _,
};

result
```

Reply:
Ooh, yeah that makes sense!